README.md
# spring-easing

[![Open Bundle](https://bundlejs.com/badge-light.svg)](https://bundlejs.com/?q=spring-easing&bundle "Check the total bundle size of spring-easing with whichever animation library you choose.")

<!-- ![](https://deno.bundlejs.com/badge?q=spring-easing) -->

[NPM](https://www.npmjs.com/package/spring-easing) <span style="padding-inline: 1rem">|</span> [Deno](https://deno.land/x/spring_easing) <span style="padding-inline: 1rem">|</span> [GitHub](https://github.com/okikio/spring-easing#readme) <span style="padding-inline: 1rem">|</span> [Docs](https://spring-easing.okikio.dev) <span style="padding-inline: 1rem">|</span> [License](#license)

Quick and easy spring animations. Works with other animation libraries ([gsap](https://greensock.com/), [animejs](https://animejs.com/), [@okikio/animate](http://npmjs.com/@okikio/animate), [motion one](https://motion.dev/), [framer motion](https://www.framer.com/docs/animation/), etc...) or the [Web Animation API (WAAPI)](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API/Using_the_Web_Animations_API), you can learn more in the [Usage](#use-with-animation-libraries) section.

`spring-easing` works by generating arrays of `frame`'s which when placed in linear order creates a smooth spring like animation.

> A `frame` represent a single frame of an animation

> _**Note**: the `spring-easing` package also supports 4 extra variants of [`spring`](https://spring-easing.okikio.dev/functions/SpringFrame), namely [`spring-in`](https://spring-easing.okikio.dev/functions/SpringInFrame), [`spring-out`](https://spring-easing.okikio.dev/functions/SpringOutFrame), [`spring-out-in`](https://spring-easing.okikio.dev/functions/SpringOutInFrame), and [`spring-in-out`](https://spring-easing.okikio.dev/functions/SpringInOutFrame), you can use these easing to create some really unique spring like animations._

<!-- > You can also read the [blog post](https://blog.okikio.dev/spring-easing), created for it's launch. -->

You can create animation's like this with `spring-easing`,

<img src="media/assets/spring-easing-demo-video.gif" width="1920" loading="lazy" alt="A demo of the various spring-easings available" align="center" style="border-radius: 1rem; 
    aspect-ratio: auto 1920 / 899;" />

<!-- https://github.com/okikio/spring-easing/assets/17222836/3813945f-b301-4399-8383-bbab227c3f68

<video controls autoplay align="center" style="border-radius: 1rem; aspect-ratio: auto 1920 / 899;">
  <source src="media/assets/spring-easing-demo-video.mp4" type="video/mp4">
</video> -->

> _Check out the spring easing variants on [Codepen](https://codepen.io/okikio/pen/MWEMEgJ)._

> _**Attention**: This entire library is a lightweight version of the [`CustomEasing`](https://native.okikio.dev/animate/api/custom-easing/) implemented in [`@okikio/animate`](https://native.okikio.dev/animate), which supports only string and number interpolation. If you'd like the complete `CustomEasing` with color interpolation, complex value interpolation, and more, go through the source code as a [Github Gist](https://gist.github.com/okikio/bed53ed621cb7f60e9a8b1ef92897471#file-custom-easing-ts), which is licensed under the MIT license._

## Installation

### Node

```bash
npm install spring-easing
```

<details>
  <summary>Others</summary>

```bash
yarn add spring-easing
```

or

```bash
pnpm install spring-easing
```

</details>

### Deno

```ts
import { SpringEasing } from "https://deno.land/x/spring_easing/mod.ts";
```

## Usage

```ts
import { SpringEasing } from "spring-easing";
// or
import SpringEasing from "spring-easing";
```

You can also use it directly through a script tag:

```html
<script src="https://unpkg.com/spring-easing" type="module"></script>
<script type="module">
  // You can then use it like this
  const { SpringEasing } = window.SpringEasing;
</script>
```

You can also use it via a CDN, e.g.

```ts
import { SpringEasing } from "https://esm.run/spring-easing";
// or
import { SpringEasing } from "https://esm.sh/spring-easing";
// or
import { SpringEasing } from "https://unpkg.com/spring-easing";
// or
import { SpringEasing } from "https://cdn.skypack.dev/spring-easing";
// or
import { SpringEasing } from "https://deno.bundlejs.com/file?q=spring-easing";
// or any number of other CDN's
```

### Use with Animation Libraries

> _**Note:** I cannot guarantee that every animation library works with `spring-easing`, for example, if an animation library doesn't support array values as keyframes, it won't work well with `spring-easing`._

The libraries that have been tested are:

| Animation Library                                                                                | Support                                                                                                                                                       | Demo                                             |
| ------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------ |
| [GSAP](https://greensock.com/)                                                                   | ✅ Yes - [Wrap Method](<https://greensock.com/docs/v3/GSAP/UtilityMethods/wrap()>)                                                                            | [Codepen](https://codepen.io/okikio/pen/MWEMEgJ) |
| [animejs](https://animejs.com)                                                                   | ✅ Yes - [Array Keyframes](https://animejs.com/documentation/#animationKeyframes)                                                                             | [Codepen](https://codepen.io/okikio/pen/MWEMEgJ) |
| [Framer Motion](https://www.framer.com/motion/)                                                  | ✅ Yes - [Array Keyframes](https://www.framer.com/docs/animation/##keyframes)                                                                                 | [Codepen](https://codepen.io/okikio/pen/MWEMEgJ) |
| [Motion One](https://motion.dev)                                                                 | ✅ Yes - [Array Keyframes](https://motion.dev/dom/animate#keyframes)                                                                                          | [Codepen](https://codepen.io/okikio/pen/MWEMEgJ) |
| [@okikio/animate](https://okikio.github.io/native/packages/animate)                              | ✅ Yes - [Array Keyframes](https://okikio.github.io/native/packages/animate/#animations)                                                                      | [Codepen](https://codepen.io/okikio/pen/MWEMEgJ) |
| [Web Animation API (WAAPI)](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API) | ✅ Yes - [Array Keyframes](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API/Keyframe_Formats#:~:text=An-,object,-containing%20key%2Dvalue) | [Codepen](https://codepen.io/okikio/pen/MWEMEgJ) |
| [Linear Easing (CSS)](https://linear-easing-generator.netlify.app)                               | ✅ Yes - [CSS Spring Easing](#CSSSpringEasing)                                                                                                                | [Codepen](https://codepen.io/okikio/pen/MWEMEgJ) |
| [Linear Easing (WAAPI)](https://linear-easing-generator.netlify.app)                             | ✅ Yes - [CSS Spring Easing](#CSSSpringEasing)                                                                                                                | [Codepen](https://codepen.io/okikio/pen/MWEMEgJ) |

e.g.

```ts
import anime from "animejs";
import { SpringEasing, SpringOutFrame } from "spring-easing";

// Note: this is the return value of `SpringEasing` and `GenerateSpringFrames`
let [translateX, duration] = SpringEasing([0, 250], {
  easing: "spring-out-in(1, 100, 10, 0)",
  // You can change the size of Array for the SpringEasing function to generate
  numPoints: 200,
  // The number of decimal places to round, final values in the generated Array
  // This option doesn't exist on `GenerateSpringFrames`
  decimal: 5,
});

anime({
  targets: "div",

  // Using spring easing animate from [0 to 250] using `spring-out-in`
  translateX,

  // You can interpolate between strings
  // You can set the easing without an easing options object
  // You can interpolate between more than 2 values
  // Remember the `0` index of `SpringEasing` is an array of spring animation keyframes
  rotate: SpringEasing(
    ["0turn", 1, 0, 0.5],
    [SpringOutFrame, 1, 100, 10, 0]
  )[0],

  // TIP... Use linear easing for the proper springy effect
  easing: "linear",

  // The optimal duration for this specific spring configuration, e.g. mass, velocity, damping, etc...
  duration,
});
```

Or

```ts
import { CSSSpringEasing } from "spring-easing";

let [easings, duration] = CSSSpringEasing({
  easing: "spring-out-in(1, 100, 10, 0)",
  // The number of decimal places to round to
  decimal: 5,
});

document.querySelector('div').animate({
  // Using spring easing animate from [0 to 250] using `spring-out-in`
  translate: ["0", "250px"],

  // You can interpolate between strings
  // You can set the easing without an easing options object
  // You can interpolate between more than 2 values
  // Remember the `0` index of `SpringEasing` is an array of spring animation keyframes
  rotate: ["0turn", "1turn", "0turn", "0.5turn"],
}, {
  // TIP... Use linear easing for the proper springy effect
  easing: `linear(${easings})`,

  // The optimal duration for this specific spring configuration, e.g. mass, velocity, damping, etc...
  duration,
});
```

> _**Note**: make sure to read the comments above, as they are valuable resources for understanding what is happening._

> Check out this demo on [Codepen](https://codepen.io/okikio/pen/MWEdzNg)

## Showcase

A couple sites/projects that use `spring-easing`:

- [`postcss-spring-easing`](https://github.com/okikio/postcss-spring-easing)
- Your site/project here...

## API

<details open>
<summary><strong><em>What's New...</em></strong></summary>

> **`NEW`** CSS Spring Easing & support for the `linear()` easing function
>
> ### CSSSpringEasing
>
> Generates a string that represents a set of values used with the linear-easing function to replicate spring animations,
> you can check out the linear-easing playground here https://linear-easing-generator.netlify.app/
> Or check out a demo on Codepen https://codepen.io/okikio/pen/vYVaEXM
>
> CSS Spring Easing has 4 properties they are `easing` (all spring frame functions are supported), `numPoints` (the size of the Array the frmae function should create), `decimal` (the number of decimal places of the values within said Array) and `quality` (how detailed/smooth the spring easing should be)..
>
> | Properties  | Default Value           |
> | ----------- | ----------------------- |
> | `easing`    | `spring(1, 100, 10, 0)` |
> | `numPoints` | `50`                    |
> | `decimal`   | `3`                     |
> | `quality`   | `0.85`                  |
>
> `CSSSpringEasing` is meant to be used with the `linear-easing()` function to replicate spring animations.
> It is based on the work done by [Jake Archibald](https://github.com/jakearchibald) in his [Linear Easing Generator](https://github.com/jakearchibald/linear-easing-generator).
>

CHANGELOG.md
## [2.3.3](https://github.com/okikio/spring-easing/compare/v2.3.2...v2.3.3) (2023-05-19)


### Bug Fixes

* **docs:** fix spring-easing deno link ([2d6b527](https://github.com/okikio/spring-easing/commit/2d6b52780bec5d822bf8e4d24f64bcc046ac5fa8))


### Misc

* **gh-bot:** :rocket: build types, api & library files ([fe1cdce](https://github.com/okikio/spring-easing/commit/fe1cdce920bdbd91827f94fc9bbe0db9e180beac))
* **gh-bot:** :rocket: release ([9904633](https://github.com/okikio/spring-easing/commit/99046331759a8b69f4e710523d0047b7289dfd50))

## [2.3.2](https://github.com/okikio/spring-easing/compare/v2.3.1...v2.3.2) (2023-05-19)


### Bug Fixes

* publish changes ([32395a4](https://github.com/okikio/spring-easing/commit/32395a462f48d359304344087a37c391b30b03ed))


### Misc

* add links to deno [skip ci] ([e92c170](https://github.com/okikio/spring-easing/commit/e92c170f6ba099a51e6485f8175d2582af0fe393))
* fix broken links & add CSSSpringEasing to demo [skip ci] ([02eb310](https://github.com/okikio/spring-easing/commit/02eb3105a22133df7b1a1dc422313fc28669f13b))
* fix format issues in package.json ([bcd3581](https://github.com/okikio/spring-easing/commit/bcd35817ea0219f2acdbb8bf4f6bf526c3d627b4))
* **gh-bot:** :rocket: build types, api & library files ([d252bec](https://github.com/okikio/spring-easing/commit/d252bec9790117c151e7043aa02889f06d577610))
* **gh-bot:** :rocket: build types, api & library files ([9e4713f](https://github.com/okikio/spring-easing/commit/9e4713fe6169b48bb8dacacba01194e2eb367e62))
* **gh-bot:** :rocket: release ([0af56b5](https://github.com/okikio/spring-easing/commit/0af56b5948e3f850d3b4e0b9832501e47b101482))
* npm keywords should be lowercase ([b053491](https://github.com/okikio/spring-easing/commit/b05349132d769a96920ed326a6c5f3d4cff01b8d))
* update deps ([0d31b4f](https://github.com/okikio/spring-easing/commit/0d31b4fc3b333e15bb5588669f21adcb84e5f334))

## [2.3.1](https://github.com/okikio/spring-easing/compare/v2.3.0...v2.3.1) (2023-05-14)


### Bug Fixes

* deno not showing docs ([6d260c4](https://github.com/okikio/spring-easing/commit/6d260c4b298e49829d2026eb00a1ffc6ce8ca541))
* git syslink issue with src/README.md [skip ci] ([42df16d](https://github.com/okikio/spring-easing/commit/42df16d8881d914d9403c6a38938a289f932aff9))

# [2.3.0](https://github.com/okikio/spring-easing/compare/v2.2.0...v2.3.0) (2023-05-14)


### Features

* add BatchSpringEasing & it's interpolation functions ([f14b603](https://github.com/okikio/spring-easing/commit/f14b603a7ced20c719c3c52d040b3d55e80aa0ea))

# [2.2.0](https://github.com/okikio/spring-easing/compare/v2.1.2...v2.2.0) (2023-05-13)


### Features

* add css spring easing and linear easing function support ([ac761fe](https://github.com/okikio/spring-easing/commit/ac761fe20c0878f3ba1966f93f816f039ff08a62))

## [2.1.2](https://github.com/okikio/spring-easing/compare/v2.1.1...v2.1.2) (2023-03-29)


### Bug Fixes

* skypack issues ([5f239d3](https://github.com/okikio/spring-easing/commit/5f239d34b3ada9d44d0db394192ab6b2cf925a17))

## [2.1.1](https://github.com/okikio/spring-easing/compare/v2.1.0...v2.1.1) (2023-03-20)


### Bug Fixes

* ... ([5429f07](https://github.com/okikio/spring-easing/commit/5429f07d0d0a14a31af287b8e6bd1479249183f1))


### Performance Improvements

* optimize perf. w/ help from [@jakearchibald](https://github.com/jakearchibald) ([1f3fc56](https://github.com/okikio/spring-easing/commit/1f3fc56ee45e8d8f9dd39c7f2fa01673c0794750))


### Reverts

* back to old interpolation syntax, as the new batch syntax causes confusion ([502bbd0](https://github.com/okikio/spring-easing/commit/502bbd0b3caa4110664a89d8163ad968253c6290))

# [2.1.0](https://github.com/okikio/spring-easing/compare/v2.0.0...v2.1.0) (2022-09-18)


### Features

* re-add instant interpolation functions ([9c52a1b](https://github.com/okikio/spring-easing/commit/9c52a1bbb0dbb1625cd8bfea46cc583d33eba59b))

# [2.0.0](https://github.com/okikio/spring-easing/compare/v1.2.0...v2.0.0) (2022-09-09)


* feat!: add `toAnimationFrames` function ([517d07e](https://github.com/okikio/spring-easing/commit/517d07efd9c7a519591f59203beae6083291db2d)), closes [#4](https://github.com/okikio/spring-easing/issues/4)


### BREAKING CHANGES

* new interpolation function syntax will break the current custom interpolation functions

# [1.2.0](https://github.com/okikio/spring-easing/compare/v1.1.1...v1.2.0) (2022-09-05)


### Bug Fixes

* bugs ([7c2e24a](https://github.com/okikio/spring-easing/commit/7c2e24a4f16f41b24ed2f201de38b8b64a66d0a7))


### Features

* add registerEasingFunction(s) function ([e97cb1d](https://github.com/okikio/spring-easing/commit/e97cb1d838f01678d86ea6bf372eaf75137446b9))


### Performance Improvements

* optimize interploation functions ([7787db5](https://github.com/okikio/spring-easing/commit/7787db509b0cf94b25ed611f23b5f0bb3298cb00))

## [1.1.1](https://github.com/okikio/spring-easing/compare/v1.1.0...v1.1.1) (2022-06-02)


### Performance Improvements

* faster interpolation for t=0 or 1  ([efbdab9](https://github.com/okikio/spring-easing/commit/efbdab9479bff49480b14909c5ff25be67c0cb8a)), closes [#3](https://github.com/okikio/spring-easing/issues/3)

# [1.1.0](https://github.com/okikio/spring-easing/compare/v1.0.0...v1.1.0) (2022-02-23)


### Bug Fixes

* broken icons ([c36509d](https://github.com/okikio/spring-easing/commit/c36509d91dec13d34bacfaa01e58ffa6ecbdefbc))


### Features

* add string interpolation & custom interpolation functions ([d13f252](https://github.com/okikio/spring-easing/commit/d13f252f9a9e0b699ea762c0f90a2e042a1e57db)), closes [#4f4](https://github.com/okikio/spring-easing/issues/4f4)


lib/mod.d.ts
export * from "./utils.ts";
export * from "./batch.ts";
export * from "./optimize.ts";
export * from "./css-linear-easing.ts";
/**
 * The format to use when defining custom frame functions
 * An example of a frame function is {@link SpringFrame}
 *
 * @source Source code of `TypeFrameFunction`
 *
 * @param t time value between 0 & 1
 * @param spring-parameters
 *  - mass = mass of object
 *  - stiffness = stiffness of spring
 *  - damping = amount to dampen spring motion
 *  - velocity = initial velocity of spring
 * @param duration (optional) the maximum duration (in milliseconds) required for a spring (with its specified spring parameters) to reach a resting position. It's used to ensure the progress of all spring frames put together are smooth
 * @returns a single frame of the frame function at the time specified
 *
 * _**Note**: Be very careful of only setting some of the spring parameters, it can cause errors if you are not careful_
 */
export type TypeFrameFunction = (t: number, [mass, stiffness, damping, velocity]?: number[], duration?: number) => number;
/**
 * The type for interpolation functions which at an instant in the animation, generate the corresponding interpolated frame
 */
export type TypeInterpolationFunction = (t: number, values: any[], decimal?: number) => string | number | any;
/*!
 * Spring solver inspired by Webkit Copyright © 2016 Apple Inc. All rights reserved. https://webkit.org/demos/spring/spring.js
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
 * THE POSSIBILITY OF SUCH DAMAGE.
*/
/**
 * Generates a single frame of the spring easing at a specific time between (0 to 1) with the spring parameters given [mass, stiffness, damping, velocity]
 *
 * @param t time value between 0 & 1
 * @param spring-parameters (limit of 0.0001 to 1)
 *  - mass = mass of object
 *  - stiffness = stiffness of spring
 *  - damping = amount to dampen spring motion
 *  - velocity = initial velocity of spring
 * @param duration (optional) the maximum duration (in milliseconds) required for a spring (with its specified spring parameters) to reach a resting position. It's used to ensure the progress of all spring frames put together are smooth
 * @returns a single frame of the spring easing at the time specified
 *
 * _**Note**: Be very careful of only setting some of the spring parameters, it can cause errors if you are not careful_
 *
 * Based on [animejs](https://github.com/juliangarnier/anime/blob/3ebfd913a04f7dc59cc3d52e38275272a5a12ae6/src/index.js#L76)
 */
export declare const SpringFrame: TypeFrameFunction;
/**
 * Cache the durations at set easing parameters
 */
export declare const EasingDurationCache: Map<string, [
    number,
    number
]>;
/**
 * The threshold for an infinite loop
 *
 * @source Source code of `INFINITE_LOOP_LIMIT`
 */
export declare const INFINITE_LOOP_LIMIT = 100000;
/**
 * The spring easing function will only look smooth at certain durations, with certain parameters.
 * This functions returns the optimal duration to create a smooth springy animation based on physics

 * @param spring-parameters
 *  - mass = mass of object
 *  - stiffness = stiffness of spring
 *  - damping = amount to dampen spring motion
 *  - velocity = initial velocity of spring
 * @return [
 *  `duration` = optimal duration for spring easings,
 *  `numPoints` = optimal num. of points to represent the spring easing
 * ]
 *
 * _**Note**: Be very careful of only setting some of the spring parameters, it can cause errors if you are not careful_
 *
 * Based on a function of the same name in [animejs](https://github.com/juliangarnier/anime/blob/3ebfd913a04f7dc59cc3d52e38275272a5a12ae6/src/index.js#L100)
 * Thanks [@jakearchibald](https://gist.github.com/jakearchibald/9718d1b81fe62d1c9655de65df1a55a4) for the help optimizing this
 */
export declare function getSpringDuration([mass, stiffness, damping, velocity]?: number[]): number[];
/**
 * Creates a new frame function where each frame follows an `out` pattern
 *
 * @source Source code of `EaseOut`
 *
 * @param frame frame function (see {@link TypeFrameFunction}, to learn more about frame functions)
 * @returns A new frame function that represents the ease-out version of the frame function given as an argument
 *
 * Based off of another library, but I can't remember which. If any devs are able to find
 * the source, I'll gladily place a link to the original source here
 */
export declare function EaseOut(frame: TypeFrameFunction): TypeFrameFunction;
/**
 * Creates a new frame function where each frame follows an `in-out` pattern
 *
 * @param frame frame function (see {@link TypeFrameFunction}, to learn more about frame functions)
 * @returns A new frame function that represents the ease-in-out version of the frame function given as an argument
 *
 * Based off of another library, but I can't remember which. If any devs are able to find
 * the source, I'll gladily place a link to the original source here
 */
export declare function EaseInOut(frame: TypeFrameFunction): TypeFrameFunction;
/**
 * Creates a new frame function where each frame follows an `out-in` pattern
 *
 * @param frame frame function (see {@link TypeFrameFunction}, to learn more about frame functions)
 * @returns A new frame function that represents the ease-in-out version of the frame function given as an argument
 *
 * Based off of another library, but I can't remember which. If any devs are able to find
 * the source, I'll gladily place a link to the original source here
 */
export declare function EaseOutIn(frame: TypeFrameFunction): TypeFrameFunction;
/**
 * Alias of {@link SpringFrame},
 * e.g. "spring in" frame function
 */
export declare const SpringInFrame: TypeFrameFunction;
/**
 * "spring-out" frame function where each {@link SpringFrame} follows an ease `out` pattern
 *
 * _**Note**: Be very careful of only setting some of the spring parameters, it can cause errors if you are not careful_
 *
 * @source Source code of `SpringOutFrame`
 */
export declare const SpringOutFrame: TypeFrameFunction;
/**
 * "spring-in-out" frame function where each {@link SpringFrame} follows an ease `in-out` pattern
 *
 * _**Note**: Be very careful of only setting some of the spring parameters, it can cause errors if you are not careful_
 *
 * @source Source code of `SpringInOutFrame`
 */
export declare const SpringInOutFrame: TypeFrameFunction;
/**
 * "spring-out-in" frame function where each {@link SpringFrame} follows an ease `out-in` pattern
 *
 * _**Note**: Be very careful of only setting some of the spring parameters, it can cause errors if you are not careful_
 *
 * @source Source code of `SpringOutInFrame`
 */
export declare const SpringOutInFrame: TypeFrameFunction;
/**
 * Given an Array of numbers, estimate the resulting number, at a `t` value between 0 to 1

 * Basic interpolation works by scaling `t` from 0 - 1, to some start number and end number, in this case lets use
 * 0 as our start number and 100 as our end number, so, basic interpolation would interpolate between 0 to 100.

 * If we use a `t` of 0.5, the interpolated value between 0 to 100, is 50.
 * {@link interpolateNumber} takes it further, by allowing you to interpolate with more than 2 values,
 * it allows for multiple values.
 * E.g. Given an Array of values [0, 100, 0], and a `t` of 0.5, the interpolated value would become 100.

 * Based on d3.interpolateBasis [https://github.com/d3/d3-interpolate#interpolateBasis],
 * check out the link above for more detail.
 *
 * @param t A number between 0 to 1
 * @param values Array of numbers to interpolate between
 * @param decimal How many decimals should the interpolated value have
 * @return Interpolated number at instant `t`
 *
 * @source Source code of `interpolateNumber`
*/
export declare function interpolateNumber(t: number, values: number[], decimal?: number): number;
/**
 * Given an Array of values, find a value using `t` (which goes from 0 to 1), by
 * using `t` to estimate the index of said value in the array of `values`

 * This is meant for interplolating strings that aren't number-like

 * @param t A number between 0 to 1
 * @param values Array of numbers to interpolate between
 * @return Interpolated numbers at different instances

 * @source Source code of `interpolateSequence`
*/
export declare function interpolateSequence<T>(t: number, values: T[]): T;
/**
 * Alias of {@link interpolateSequence}
 * @deprecated please use {@link interpolateSequence}, it's the same functionality but different name
 */
export declare const interpolateUsingIndex: typeof interpolateSequence;
/**
 * Functions the same way {@link interpolateNumber} works.
 * Convert strings to numbers, and then interpolates the numbers,
 * at the end if there are units on the first value in the `values` array,
 * it will use that unit for the interpolated result.
 * Make sure to read {@link interpolateNumber}.
 *
 * @source Source code of `interpolateString`
*/
export declare function interpolateString(t: number, values: (string | number)[], decimal?: number): string;
/**
 * Interpolates all types of values including number, and string values.
 * Make sure to read {@link interpolateNumber}, {@link interpolateString} and {@link interpolateSequence}.
 *
 * @source Source code of `interpolateComplex`
*/
export declare function interpolateComplex<T>(t: number, values: T[], decimal?: number): string | number | T;
/**
 * The array frame function format for easings,
 * @example
 * `[SpringFrame, mass, stiffness, damping, velocity]`
 */
export type TypeArrayFrameFunctionFormat = [TypeFrameFunction, ...number[]];
/**
 * The list of spring easing functions
 */
export declare let EasingFunctions: {
    spring: TypeFrameFunction;
    "spring-in": TypeFrameFunction;
    "spring-out": TypeFrameFunction;
    "spring-in-out": TypeFrameFunction;
    "spring-out-in": TypeFrameFunction;
};
export declare let EasingFunctionKeys: string[];
/**
 * Allows you to register new easing functions
 */
export declare function registerEasingFunction<T extends string>(key: T, fn?: TypeFrameFunction): void;
/**
 * Allows you to register multiple new easing functions
 */
export declare function registerEasingFunctions<T extends Record<string, TypeFrameFunction>>(obj: T): void;
/**
 * The formats supported for easings,
 * @example
 * * `"spring(mass, stiffness, damping, velocity)"`
 * * `"spring-in(mass, stiffness, damping, velocity)"`
 * * `"spring-out(mass, stiffness, damping, velocity)"`
 * * `"spring-in-out(mass, stiffness, damping, velocity)"`
 * * `"spring-out-in(mass, stiffness, damping, velocity)"`
 * `[SpringFrame, mass, stiffness, damping, velocity]`
 */
export type TypeEasings = `${keyof typeof EasingFunctions}` | `${keyof typeof EasingFunctions}(${string})` | (string & {}) | TypeArrayFrameFunctionFormat;
/**
 * Spring Easing has 3 properties they are `easing` (all spring frame functions are supported), `numPoints` (the size of the Array the frmae function should create), and `decimal` (the number of decimal places of the values within said Array).
 *
 * | Properties  | Default Value           |
 * | ----------- | ----------------------- |
 * | `easing`    | `spring(1, 100, 10, 0)` |
 * | `numPoints` | `50`                    |
 * | `decimal`   | `3`                     |
 */
export type TypeEasingOptions = {
    /**
     * By default, Spring Easing support easings in the form,
     *
     * | constant   | accelerate         | decelerate     | accelerate-decelerate | decelerate-accelerate |
     * | :--------- | :----------------- | :------------- | :-------------------- | :-------------------- |
     * |            | spring / spring-in | spring-out     | spring-in-out         | spring-out-in         |
     *
     * All **Spring** easing's can be configured using theses parameters,
     *
     * `"spring-*(mass, stiffness, damping, velocity)"` or
     * `[SpringOutFrame, mass, stiffness, damping, velocity]`
     *
     * Each parameter comes with these defaults
     *
     * | Parameter | Default Value |
     * | --------- | ------------- |
     * | mass      | `1`           |
     * | stiffness | `100`         |
     * | damping   | `10`          |
     * | velocity  | `0`           |
     */
    easing?: TypeEasings;
    numPoints?: number;
    decimal?: number;
};
/**
 * Convert easing parameters to Array of numbers, e.g. "spring(2, 500)" to [2, 500]
 *
 * Based off of [animejs](https://github.com/juliangarnier/anime/blob/3ebfd913a04f7dc59cc3d52e38275272a5a12ae6/src/index.js#L69)
 */
export declare function parseEasingParameters(str: string): (string | number)[];
/**
 * Returns a {@link TypeEasingOptions} object from a easing "string", or frame function array
 *
 * _**Note**: Be very careful of only setting some of the spring parameters, it can cause errors if you are not careful_
 */
export declare function EasingOptions<T extends TypeEasingOptions>(options?: T | TypeEasingOptions["easing"]): {
    easing: TypeArrayFrameFunctionFormat;
    numPoints: number;
    decimal: number;
} & Omit<T, "easing" | "numPoints" | "decimal">;
/**
 * Cache generated frame points for commonly used easing functions
 */
export declare const FramePtsCache: Map<string, WeakMap<Function, [number[], number]>>;
/**
 * Create an Array of frames using the easing specified.
 * The array size `numPoints` large, which is by default 100.
 * Easing can be use custom defined frame functions, so, instead of string you can use,
 * e.g.
 * ```ts
 * GenerateSpringFrames({
 *     easing: [SpringOutInFrame, 1, 100, 10, 0],
 *     numPoints: 100
 * })
 * ```
 *
 * _**Note**: Be very careful of only setting some of the spring parameters, it can cause errors if you are not careful_
 *
 * @param options Accepts {@link TypeEasingOptions EasingOptions}
 *
 * Based on https://github.com/w3c/csswg-drafts/issues/229#issuecomment-861415901
 */
export declare function GenerateSpringFrames(options?: TypeEasingOptions): [number[], number];
/**
 * Generates an Array of values using frame functions which in turn create the effect of spring easing.
 * To use this properly make sure to set the easing animation option to "linear".
 * Check out a demo of SpringEasing at <https://codepen.io/okikio/pen/MWEdzNg>
 *
 * SpringEasing has 3 properties they are `easing` (all the easings from {@link EasingFunctions} are supported on top of frame functions like SpringFrame, SpringFrameOut, etc..), `numPoints` (the size of the Array the frame function should create), and `decimal` (the number of decimal places of the values within said Array).
 *
 * | Properties  | Default Value           |
 * | ----------- | ----------------------- |
 * | `easing`    | `spring(1, 100, 10, 0)` |
 * | `numPoints` | `50`                    |
 * | `decimal`   | `3`                     |
 *
 * By default, Spring Easing support easings in the form,
 *
 * | constant   | accelerate         | decelerate     | accelerate-decelerate | decelerate-accelerate |
 * | :--------- | :----------------- | :------------- | :-------------------- | :-------------------- |
 * |            | spring / spring-in | spring-out     | spring-in-out         | spring-out-in         |
 *
 * All **Spring** easing's can be configured using theses parameters,
 *
 * `spring-*(mass, stiffness, damping, velocity)`
 *
 * Each parameter comes with these defaults
 *
 * | Parameter | Default Value |
 * | --------- | ------------- |
 * | mass      | `1`           |
 * | stiffness | `100`         |
 * | damping   | `10`          |
 * | velocity  | `0`           |
 *
 *  e.g.
 *  ```ts
 *  import { SpringEasing, SpringOutFrame } from "spring-easing";
 *  import anime from "animejs";
 *
 *  // Note: this is the return value of {@link SpringEasing} and {@link GenerateSpringFrames}, you don't need the object to get this format
 *  let [translateX, duration] = SpringEasing([0, 250], {
 *      easing: "spring-out-in(1, 100, 10, 0)",
 *
 *      // You can change the size of Array for the SpringEasing function to generate
 *      numPoints: 200,
 *
 *      // The number of decimal places to round, final values in the generated Array
 *      // This option doesn't exist on {@link GenerateSpringFrames}
 *      decimal: 5,
 *  });
 *
 *  anime({
 *    targets: "div",
 *
 *    // Using spring easing animate from [0 to 250] using `spring-out-in`
 *    translateX,
 *
 *    // You can set the easing without an object
 *    rotate: SpringEasing(["0turn", 1, 0, 0.5], [SpringOutFrame, 1, 100, 10, 0])[0],
 *
 *    // TIP... Use linear easing for the proper effect
 *    easing: "linear",
 *
 *    // The optimal duration for this specific spring
 *    duration
 *  })
 *  ```
 *
 * @param values Values to animate between, e.g. `["50px", 60]`
 * > _**Note**: You can interpolate with more than 2 values, but it's very confusing, so, it's best to choose 2_
 * @param options Accepts {@link TypeEasingOptions EasingOptions} or {@link TypeEasingOptions["easing"] array frame functions}
 * @param interpolationFunction If you wish to use your own interpolation functions you may
 * @return
 * ```ts
 * // An array of keyframes that represent said spring animation and
 * // Total duration (in milliseconds) required to create a smooth spring animation
 * [
 *   [50, 55, 60, 70, 80, ...],
 *   3500
 * ]
 * ```
 */
export declare function SpringEasing<T>(values: T[], options?: TypeEasingOptions | TypeEasingOptions["easing"], customInterpolate?: TypeInterpolationFunction): readonly [any[], number];
export default SpringEasing;
//# sourceMappingURL=mod.d.ts.map